File: packages/vision/scene_classifier.py
"""
Scene-level safety classifier.

Answers "what is happening in this image" rather than "what objects are
present" — the object/scene split in SKILL.md §9. Object detection can only
flag things it has a box for; a violent or graphic scene has no box to draw,
so it produces whole-frame protection instead.

Model: `image-safety-classifier-xs` (SwiftFormer-XS finetune, MIT).
Selected in Phase A — see docs/06-ai-detection.md for the benchmark and the
reasoning. Outputs three probabilities: NSFL (gore/violence), NSFW (sexual),
SFW.

This class deliberately does not subclass VisionModel: that interface
returns `List[Detection]` (objects with boxes), which a scene classifier
has none of. Forcing it to fit would mean inventing fake bounding boxes,
which SKILL.md §13 explicitly forbids.
"""
import logging
import time
from pathlib import Path
from typing import Optional

import numpy as np

from packages.shared.config import MyKidConfig
from packages.shared.errors import InferenceError, ModelError
from packages.shared.logger import LogEvent, get_logger, log_event
from packages.shared.types import SceneRisk


logger = get_logger("mykid.vision.scene_classifier")


# Fixed by the model's own config (pretrained_cfg.label_names) — read from
# the model, not assumed.
SCENE_LABELS = ["NSFL", "NSFW", "SFW"]

MODEL_INPUT_SIZE = 224


class SceneClassifier:
    """
    Classifies an image's scene-level safety risk.

    Produces a SceneRisk, which the risk engine compares against
    `scene_risk_threshold` to decide whole-frame protection.
    """

    def __init__(self, config: MyKidConfig, model_path: Optional[str] = None):
        """
        Args:
            config: Global configuration.
            model_path: Override for the classifier weights. Defaults to
                `models/candidates/image-safety-classifier-xs.onnx`.
        """
        self.config = config
        self.model_path = Path(
            model_path
            or Path(config.model.path) / "candidates" / "image-safety-classifier-xs.onnx"
        )
        self.session = None

    def load(self) -> None:
        """
        Load the ONNX classifier.

        Raises:
            ModelError: If the runtime is missing or the weights can't load.
        """
        try:
            import onnxruntime as ort
        except ImportError as exc:
            raise ModelError(
                "onnxruntime is not installed. Run 'pip install -r requirements.txt'"
            ) from exc

        if not self.model_path.exists():
            raise ModelError(
                f"Scene classifier weights not found at {self.model_path}. "
                "Run 'python scripts/benchmark_safety_models.py' to download them."
            )

        try:
            self.session = ort.InferenceSession(
                str(self.model_path), providers=["CPUExecutionProvider"]
            )
        except Exception as exc:
            error_msg = f"Failed to load scene classifier: {exc}"
            log_event(
                logger, logging.ERROR, LogEvent.ERROR_MODEL, error_msg,
                metadata={"error": str(exc)},
            )
            raise ModelError(error_msg) from exc

        log_event(
            logger,
            logging.INFO,
            LogEvent.MODEL_LOADED,
            "Scene classifier loaded",
            metadata={"model": self.model_path.name, "labels": SCENE_LABELS},
        )

    def _preprocess(self, image: np.ndarray) -> np.ndarray:
        """
        Resize to 224x224 and arrange as NCHW float32 in the raw 0-255 range.

        Normalisation is baked into the ONNX graph, so the input must NOT be
        pre-normalised. Verified in Phase A: normalising first pushes the
        NSFW score on a photo of a bus from 0.032 to 0.18, because the graph
        then normalises already-normalised values.
        """
        import cv2

        resized = cv2.resize(
            image, (MODEL_INPUT_SIZE, MODEL_INPUT_SIZE), interpolation=cv2.INTER_LINEAR
        )
        chw = np.transpose(resized.astype(np.float32), (2, 0, 1))
        return np.expand_dims(chw, axis=0)

    def predict(self, image: np.ndarray) -> SceneRisk:
        """
        Classify an RGB image's scene-level risk.

        Args:
            image: RGB numpy array (H, W, 3).

        Returns:
            SceneRisk with `graphic` and `sexual` populated. `violence` is
            left at 0.0: this model folds violence into its NSFL class
            rather than scoring it separately, and inventing a number for it
            would misrepresent what was measured.

        Raises:
            InferenceError: If inference fails.
        """
        if self.session is None:
            raise ModelError("Scene classifier not loaded. Call load() first.")

        if not isinstance(image, np.ndarray) or image.size == 0:
            raise InferenceError("Invalid or empty image array")

        try:
            start = time.perf_counter()
            probabilities = self.session.run(
                None, {"image": self._preprocess(image)}
            )[0][0]
            elapsed_ms = (time.perf_counter() - start) * 1000
        except Exception as exc:
            error_msg = f"Scene classification failed: {exc}"
            log_event(
                logger, logging.ERROR, LogEvent.ERROR_INFERENCE, error_msg,
                metadata={"error": str(exc)},
            )
            raise InferenceError(error_msg) from exc

        scores = dict(zip(SCENE_LABELS, (float(p) for p in probabilities)))
        scene_risk = SceneRisk(graphic=scores["NSFL"], sexual=scores["NSFW"])

        if scene_risk.max_score() >= self.config.detection.scene_risk_threshold:
            log_event(
                logger,
                logging.INFO,
                LogEvent.RISK_DETECTED,
                "Scene-level risk detected",
                metadata={
                    "graphic": round(scene_risk.graphic, 4),
                    "sexual": round(scene_risk.sexual, 4),
                    "elapsed_ms": round(elapsed_ms, 1),
                },
            )

        return scene_risk

File: apps/extension/src/inference/scene-classifier.js
/**
 * MyKid — Scene-level safety classifier (Phase B)
 *
 * Answers "what is happening in this image" rather than "what objects are
 * present" (SKILL.md §9). This is what actually detects the content the
 * product exists for — gore, violence, sexual content — none of which the
 * COCO object detector has any concept of.
 *
 * Model: `image-safety-classifier-xs` (SwiftFormer-XS finetune, MIT),
 * selected in Phase A. See docs/06-ai-detection.md for the benchmark.
 *
 * Scene risk has no bounding boxes by nature, so it produces whole-frame
 * protection. Inventing boxes for it is explicitly forbidden by
 * SKILL.md §13.
 *
 * The browser mirror of packages/vision/scene_classifier.py — keep the two
 * in step; Phase 13 checks they agree.
 *
 * Exposes a global `MyKidSceneClassifier`.
 */
var MyKidSceneClassifier = (function () {
  const INPUT_SIZE = 224;

  // Fixed by the model's own config (pretrained_cfg.label_names), read from
  // the model rather than assumed.
  const LABELS = ["NSFL", "NSFW", "SFW"];

  /**
   * Resize to 224x224 as raw 0-255 RGB in NCHW order.
   *
   * Normalisation is baked into the ONNX graph, so the input must NOT be
   * pre-normalised. Verified in Phase A: normalising first pushes the NSFW
   * score on a photo of a bus from 0.032 to 0.18, because the graph then
   * normalises already-normalised values.
   *
   * Note this is a plain resize, not the letterbox used for detection —
   * a classifier has no boxes to map back, so aspect distortion costs
   * nothing and matches how the model was trained.
   */
  function preprocess(source) {
    const canvas = new OffscreenCanvas(INPUT_SIZE, INPUT_SIZE);
    const ctx = canvas.getContext("2d", { willReadFrequently: true });
    ctx.drawImage(source, 0, 0, INPUT_SIZE, INPUT_SIZE);

    return packPixels(ctx.getImageData(0, 0, INPUT_SIZE, INPUT_SIZE).data);
  }

  /**
   * RGBA bytes -> NCHW float32, alpha dropped, values left in 0-255.
   *
   * Split out from preprocess() so it can be tested without a canvas: this
   * is where a silent divergence from the Python pipeline would live. Get
   * the channel order or the layout wrong and the model still runs, still
   * returns three plausible probabilities, and is quietly wrong — which is
   * exactly the failure a test has to catch. See
   * tests/integration/test_scene_parity.py.
   */
  function packPixels(rgba) {
    const pixelCount = INPUT_SIZE * INPUT_SIZE;
    const tensor = new Float32Array(pixelCount * 3);

    const greenOffset = pixelCount;
    const blueOffset = pixelCount * 2;

    for (let i = 0; i < pixelCount; i++) {
      const offset = i * 4;
      tensor[i] = rgba[offset];
      tensor[greenOffset + i] = rgba[offset + 1];
      tensor[blueOffset + i] = rgba[offset + 2];
    }

    return tensor;
  }

  /**
   * Classify one image's scene-level risk.
   *
   * @param {ort.InferenceSession} session
   * @param {ImageBitmap} source
   * @returns {Promise<{sceneRisk: Object, scores: Object, inferMs: number}>}
   */
  async function classify(session, source) {
    const tensorData = preprocess(source);
    const inputTensor = new ort.Tensor("float32", tensorData, [
      1, 3, INPUT_SIZE, INPUT_SIZE,
    ]);

    const start = performance.now();
    const outputMap = await session.run({ [session.inputNames[0]]: inputTensor });
    const probabilities = outputMap[session.outputNames[0]].data;
    const inferMs = performance.now() - start;

    const scores = {};
    LABELS.forEach((label, index) => {
      scores[label] = probabilities[index];
    });

    return {
      // `violence` stays 0: this model folds violence into NSFL rather
      // than scoring it separately, and putting a number there we didn't
      // measure would misrepresent the model. The field is kept so a
      // dedicated violence model can fill it later.
      sceneRisk: {
        violence: 0,
        graphic: scores.NSFL,
        sexual: scores.NSFW,
      },
      scores,
      inferMs: Math.round(inferMs),
    };
  }

  return { INPUT_SIZE, LABELS, preprocess, packPixels, classify };
})();
